test/conformance/exp_enqueue_native/enqueue_native_level_zero.cpp

```diff
@@ -0,0 +1,128 @@
+    InteropData1 *func_data = reinterpret_cast<InteropData1 *>(data);
+
+    // If L0 interop becomes a real use case we should make a new UR entry point
+    // to propagate events into and out of the the interop func.
+    zeCommandListAppendMemoryFill(
+        CommandList, func_data->fill_ptr,
+        &urLevelZeroEnqueueNativeCommandTest::val,
+        sizeof(urLevelZeroEnqueueNativeCommandTest::val),
+        urLevelZeroEnqueueNativeCommandTest::allocation_size, nullptr, 0,
+        nullptr);
+}
+
+struct InteropData2 {
+    void *from, *to;
+};
+
+// Read from device ptr to host ptr
+void interop_func_2(ur_queue_handle_t hQueue, void *data) {
+    ze_command_list_handle_t CommandList;
+    ASSERT_SUCCESS(urQueueGetNativeHandle(hQueue, nullptr,
+                                          (ur_native_handle_t *)&CommandList));
+    InteropData2 *func_data = reinterpret_cast<InteropData2 *>(data);
+
+    // If L0 interop becomes a real use case we should make a new UR entry point
+    // to propagate events into and out of the the interop func.
+    zeCommandListAppendMemoryCopy(
+        CommandList, func_data->to, func_data->from,
+        urLevelZeroEnqueueNativeCommandTest::allocation_size, nullptr, 0,
+        nullptr);
+}
+
+TEST_P(urLevelZeroEnqueueNativeCommandTest, Success) {
+    InteropData1 data_1{device_ptr};
+    ur_event_handle_t event_1;
+    ASSERT_SUCCESS(urEnqueueNativeCommandExp(
+        queue, &interop_func_1, &data_1, 0, nullptr /*phMemList=*/,
+        nullptr /*pProperties=*/, 0, nullptr /*phEventWaitList=*/, &event_1));
+}
+
+TEST_P(urLevelZeroEnqueueNativeCommandTest, Dependencies) {
+    ur_event_handle_t event_1, event_2;
+
+    InteropData1 data_1{device_ptr};
+    ASSERT_SUCCESS(urEnqueueNativeCommandExp(
+        queue, &interop_func_1, &data_1, 0, nullptr /*phMemList=*/,
+        nullptr /*pProperties=*/, 0, nullptr /*phEventWaitList=*/, &event_1));
+
+    InteropData2 data_2{device_ptr, host_vec.data()};
+    ASSERT_SUCCESS(urEnqueueNativeCommandExp(
+        queue, &interop_func_2, &data_2, 0, nullptr /*phMemList=*/,
+        nullptr /*pProperties=*/, 1, &event_1, &event_2));
+    urQueueFinish(queue);
+    for (auto &i : host_vec) {
+        ASSERT_EQ(i, val);
+    }
+}
+
+TEST_P(urLevelZeroEnqueueNativeCommandTest, DependenciesURBefore) {
+    ur_event_handle_t event_1, event_2;
+
+    ASSERT_SUCCESS(urEnqueueUSMFill(queue, device_ptr, sizeof(val), &val,
+                                    allocation_size, 0,
+                                    nullptr /*phEventWaitList=*/, &event_1));
+
+    InteropData2 data_2{device_ptr, host_vec.data()};
+    ASSERT_SUCCESS(urEnqueueNativeCommandExp(
+        queue, &interop_func_2, &data_2, 0, nullptr /*phMemList=*/,
+        nullptr /*pProperties=*/, 1, &event_1, &event_2));
+    urQueueFinish(queue);
+    for (auto &i : host_vec) {
+        ASSERT_EQ(i, val);
+    }
+}
+
+TEST_P(urLevelZeroEnqueueNativeCommandTest, DependenciesURAfter) {
```

Comment:
this is failing.

If the queue this test is using is out of order, this may be a result of the problem we talked about. If that is indeed the case, one of the solutions might be something like:
```
if (!isInOrderQueue) {
  queueFinish();
  // queue finish will execute the command list. Open it again so that `zeCommandListAppendSignalEvent` can be executed.
  UR_CALL(Queue->Context->getAvailableCommandList(
      Queue, CommandList, UseCopyEngine, NumEventsInWaitList, phEventList));
}
```
@nrspruit do you know whether, in the case of out-of-order command list, there's a better way to ensure ordering between the three different operations (wait, callback, signal) in the implementation of `urEnqueueNativeCommandExp`?

The current one is based on [`enqueueUSMPrefetch`](https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/level_zero/memory.cpp#L1203), but I suspect the prefetch implementation is also incorrect.

Reply:
Have updated. Let me know what you think